### PR TITLE
cmd, p2p: update p2p config, add cmd protocol flag

### LINKED ISSUE #467: Connect and exchange messages on kusama
## Expected Behavior

We should be able to connect to Kusama, exchange status messages, send a block request message and receive a block response message.

## Current Behavior

We previously attempted to connect to Alexander. See #227 for more details.

## Related Issues

- [ ] #443
- [ ] #521
- [x] #552 / #538
- [ ] #554
- [ ] #563

### LINKED ISSUE #484: Implement LEB128 variable-length encoding
## Expected Behavior

We have LEB123 variable-length encoding implemented. 

## Current Behavior

We have an unused `leb128ToUint64` method that decodes a byte array to uint64 using LEB128 variable-length encoding. We previously thought we needed this to send and receive messages with Alexander. We need to investigate whether we still need this and implement it if we do.

### PR DESCRIPTION
**Pull request by ryanchristo**
## Changes

Implement variable length encoding for network messages.

Update network identity to use ed25519 and gossamer configuration to connect to kusama. "Similar to other decentralized networks, each Polkadot RE node possesses a network private key and a network public key representing an ED25519 key pair" (4.1). 

Implement `--protocol` flag to specify a protocol id for the host to use.

## Tests:

```
make gossamer
```

successfully connect and receive "host status message not set error" (#443):
```
./build/bin/gossamer init
./build/bin/gossamer --verbosity trace
```

successfully connect using custom protocol and receive "host status message not set error" (#443):
```
./build/bin/gossamer init
./build/bin/gossamer --protocol /hello/world --verbosity trace
```

receive error "json: cannot unmarshal array" (#552):
```
./build/bin/gossamer init --genesis config/ksmcc3.json
```

receive error "connected to wrong peer" (#554):
```
./build/bin/gossamer init
./build/bin/gossamer --bootnodes /dns4/p2p.cc3-0.kusama.network/tcp/30100/p2p/QmeCit3Nif4VfNqrEJsdYHZGcKzRCnZvGxg6hha1iNj4mk --verbosity trace
```

successfully connect and receive "host status message not set error" (#443):
```
./build/bin/gossamer init
./build/bin/gossamer --bootnodes /dns4/p2p.cc3-0.kusama.network/tcp/30100/p2p/12D3KooWDgtynm4S9M3m6ZZhXYu2RrWKdvkCSScc25xKDVSg1Sjd --verbosity trace
```

### Issues:

https://hackmd.io/ksLK42jcRJGjKQuSNAvAWw?view#Connection-Examples

related to #467, closes #484

### CODE REVIEW

File: p2p/utils.go [hunk 1]
```diff
@@ -71,7 +72,8 @@ func generateKey(seed int64, fp string) (crypto.PrivKey, error) {
 	} else {
 		r = mrand.New(mrand.NewSource(seed))
 	}
-	key, _, err := crypto.GenerateECDSAKeyPair(r)
+	// TODO: secp256k1 in polkadot / ed25519 in specification
```
Comment (ryanchristo) [inline p2p/utils.go]: addressed in #554

File: p2p/utils.go [hunk 2]
```diff
@@ -101,7 +103,8 @@ func loadKey(fp string) (crypto.PrivKey, error) {
 	if err != nil {
 		return nil, err
 	}
-	return crypto.UnmarshalECDSAPrivateKey(dec)
+	// TODO: secp256k1 in polkadot / ed25519 in specification
```
Comment (ryanchristo) [inline p2p/utils.go]: addressed in #554

File: p2p/host.go [hunk 3]
```diff
@@ -64,7 +65,6 @@ func newHost(ctx context.Context, cfg *Config) (*host, error) {
 		libp2p.DisableRelay(),
 		libp2p.Identity(cfg.privateKey),
 		libp2p.NATPortMap(),
-		libp2p.Ping(true),
```
Comment (ryanchristo) [inline p2p/host.go]: enabled by default

**Comment by ryanchristo** [comment]
Also, the error received when running `--genesis config/ksmcc3.json` is addressed in #552

File: p2p/host.go [hunk 4]
```diff
@@ -188,6 +188,12 @@ func (h *host) send(p peer.ID, msg Message) (err error) {
 		return err
 	}
 
+	// variable length encoding
+	_, err = s.Write(common.Uint16ToBytes(uint16(len(encMsg)))[0:1])
```
Comment (noot) [inline p2p/host.go]: is this LEB128?

**Comment by noot** [comment]
also i get a SIGABRT when I run the first set of instructions :/

File: p2p/host.go (hunk 4 above)
Comment (ryanchristo) [reply p2p/host.go]: I am not actually sure if this is officially LEB128 encoding. This is what was here before. It was commented out at some point, then removed per clean up request.

What I do know is that this is what currently works with polkadot and is the solution to the read buffer error I mentioned at some point, `b"": Not enough data to fill buffer`.

**Comment by ryanchristo** [comment]
@noot I updated the instructions and I am not able to reproduce. Hmm...

**Comment by ryanchristo** [comment]
Yes it should, I will fix.

**Comment by ryanchristo** [comment]
cc @edwardmack @noot I have updated the tests, using `ed25519` for the network identity properly now, and I have updated the examples. @noot Let me know if you are still getting the abort error.

File: cmd/gossamer/configcmd.go [hunk 5]
```diff
@@ -84,7 +84,7 @@ func makeNode(ctx *cli.Context) (*dot.Dot, *cfg.Config, error) {
 		return nil, nil, err
 	}
 
-	log.Info("🕸\t Configuring node...", "datadir", fig.Global.DataDir, "protocolID", string(gendata.ProtocolId), "bootnodes", fig.P2p.BootstrapNodes)
+	log.Info("🕸\t Configuring node...", "datadir", fig.Global.DataDir, "protocol", fig.P2p.ProtocolId, "bootnodes", fig.P2p.BootstrapNodes)
```
Comment (thomasmodeneis) [inline cmd/gossamer/configcmd.go]: should be protocolID:, raised https://github.com/ChainSafe/gossamer/issues/559 to have lint reporting things like this next time.

Comment (ryanchristo) [reply cmd/gossamer/configcmd.go]: It looks like you update this in #555 already... should it be addressed there?

Comment (ryanchristo) [reply cmd/gossamer/configcmd.go]: It looks like you update this in #555 already... should it be addressed there?

Comment (thomasmodeneis) [reply cmd/gossamer/configcmd.go]: I will resolve that in the ticket mentioned after this is merged.

**Comment by noot** [comment]
@ryanchristo i'm still getting a SIGABRT when trying to run gossamer :/ if no one else is getting this i can just do a fresh clone or something

**Comment by noot** [comment]
@ryanchristo on development I don't get a SIGABRT, I'm gonna look and see if I can tell what's causing it

File: p2p/host.go (hunk 4 above)
Comment (noot) [reply p2p/host.go]: if you use the LEB encoding function from before that would probably be preferred, for longer length messages the encoding might be >2 bytes, so this wouldn't work

**Comment by ryanchristo** [comment]
cc @noot ready for another look when you get a chance

File: p2p/utils.go [hunk 6]
```diff
@@ -135,12 +135,13 @@ func saveKey(priv crypto.PrivKey, fp string) (err error) {
...
 
-// TODO: implement LEB128 variable-length encoding
-
-// Decodes a byte array to uint64 using LEB128 variable-length encoding
-// func leb128ToUint64(in []byte) uint64 {
-// 	return leb128.ToUInt64(in)
-// }
+// encodeMessageLEB128 applies leb128 variable length encoding
+func encodeMessageLEB128(encMsg []byte) []byte {
+	n := leb128.ToUInt64(encMsg)
```
Comment (noot) [inline p2p/utils.go]: I don't think this is correct?
`ToUint64` decodes a LEB128 encoded byte array
I think you want to use `out := leb128.FromUint64(uint64(len(encMsg))); return append(out, encMsg...)`
https://godoc.org/github.com/filecoin-project/go-leb128

Comment (ryanchristo) [reply p2p/utils.go]: oof, that makes more sense

Comment (ryanchristo) [reply p2p/utils.go]: fixed!

**Pull request merged**